### DISCUSSION TITLE
torch.compile with stable diffusion

### DISCUSSION THREAD

**Discussion by Madhumitha-MCW** [huggingface/diffusers#7828] [Q&A]
I would like to know why any inference example with stable diffusion has only unet part of the model passed into torch.compile(). https://pytorch.org/TensorRT/tutorials/_rendered_examples/dynamo/torch_compile_stable_diffusion.html

1. When using torch.compile with model.vae it relatively takes longer time than just compiling unet part. So, is it recommended to compile only unet or what is the exact reason?

2. And if I want to execute in graph mode, will all other uncompiled parts undergo eager execution if I compile only unet part?

**Comment by tolgacangoz** [comment]
1. Most of the time, `torch.compile()` is applied to UNet only because it is UNet that is responsible for most of the calculations. You can compile as many components as possible such as [VAE](https://pytorch.org/blog/accelerating-generative-ai-3/#compiling-the-unet-and-vae), [ControlNet](https://huggingface.co/docs/diffusers/main/en/optimization/torch2.0#controlnet:~:text=pipe.controlnet%20%3D%20torch.compile(pipe.controlnet%2C%20mode%3D%22reduce%2Doverhead%22%2C%20fullgraph%3DTrue)), [text encoder(s)](https://huggingface.co/docs/diffusers/main/en/api/pipelines/deepfloyd_if#optimizing-for-speed:~:text=pipe.text_encoder%20%3D%20torch.compile(pipe.text_encoder%2C%20mode%3D%22reduce%2Doverhead%22%2C%20fullgraph%3DTrue)), and even [optimizer](https://pytorch.org/tutorials/recipes/compiling_optimizer.html), [sampler function](https://pytorch.org/blog/accelerated-generative-diffusion-models/#conclusions-and-next-steps:~:text=At%20the%20moment,every%20sampling%20step), [`torchvision.transforms.v2`](https://pytorch.org/vision/stable/transforms.html#transforms-perf:~:text=You%20may%20want%20to%20experiment%20a%20bit%20if%20you%E2%80%99re%20chasing%20the%20very%20best%20performance.%20Using%20torch.compile()%20on%20individual%20transforms%20may%20also%20help%20factoring%20out%20the%20memory%20format%20variable%20(e.g.%20on%20Normalize).). But the last three may not be stable and proper yet. Are you saying that compiling both UNet and VAE is slower than compiling only UNet? Speeding-up thanks to compiling is observed in relatively new GPUs more.
2. I think so.

**Comment by Madhumitha-MCW** [comment]
Yes, the time taken to compile vae is more than only compiling unet part. Thanks on your reply!

**Reply by tolgacangoz** [reply]
Are you talking about the first inference time? Don't count the first inference. Evaluate the next ones. The first inference (the compilation part) could take more time for more compilation tasks.

**Comment by Madhumitha-MCW** [comment]
Yes, I just was mentioning the first inference time(compilation part). The final inference is not slow.

**Reply by tolgacangoz** [reply]
My commonsense -> If someone compiles more parts then wouldn't it take more time to compile?

**Comment by Madhumitha-MCW** [comment]
Yes, it would. I first thought the entire inference time was longer, but when I verified it, it was just the initial part which took longer
